src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs

```diff
@@ -600,20 +610,70 @@ public override SecurityToken CreateToken(SecurityTokenDescriptor tokenDescripto
             if (tokenDescriptor == null)
                 throw LogHelper.LogArgumentNullException(nameof(tokenDescriptor));
 
+            // TODO at next major version (8.0) use only Audiences as SecurityTokenDescriptor.Audience" will be removed.
+            string aud = tokenDescriptor.Audience;
+            IDictionary<string, object> claims = tokenDescriptor.Claims;
+
+            if (!tokenDescriptor.Audiences.IsNullOrEmpty())
+            {
+                claims = MergeClaimsAudValuesAndDescriptorAudiences(tokenDescriptor);
+                aud = null;
+            }
+
             return CreateJwtSecurityTokenPrivate(
                 tokenDescriptor.Issuer,
-                tokenDescriptor.Audience,
+                aud,
                 tokenDescriptor.Subject,
                 tokenDescriptor.NotBefore,
                 tokenDescriptor.Expires,
                 tokenDescriptor.IssuedAt,
                 tokenDescriptor.SigningCredentials,
                 tokenDescriptor.EncryptingCredentials,
-                tokenDescriptor.Claims,
+                claims,
                 tokenDescriptor.TokenType,
                 tokenDescriptor.AdditionalHeaderClaims,
                 tokenDescriptor.AdditionalInnerHeaderClaims);
         }
+        private static IDictionary<string, object> MergeClaimsAudValuesAndDescriptorAudiences(SecurityTokenDescriptor tokenDescriptor)
+        {
+            if (tokenDescriptor.Claims == null)
+                return new Dictionary<string, object>(){{JwtRegisteredClaimNames.Aud, tokenDescriptor.Audiences}};
+
+            if (tokenDescriptor.Claims.ContainsKey(JwtRegisteredClaimNames.Aud))
+                MergeAudClaimsIntoAudiencesProperty(tokenDescriptor);
+
+            var claimsCopy = new Dictionary<string, object>(tokenDescriptor.Claims);
+            claimsCopy[JwtRegisteredClaimNames.Aud] = tokenDescriptor.Audiences;
+            return claimsCopy;
+        }
+
+        private static void MergeAudClaimsIntoAudiencesProperty(SecurityTokenDescriptor tokenDescriptor)
+        {
+            // This switch checks for Uri in addition to string per the definition of the 'aud' claim in the JWT RFC:
+            // https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3
+            switch (tokenDescriptor.Claims[JwtRegisteredClaimNames.Aud])
```

Comment:
This code modifies a SecurityTokenDescriptor that is owned by user, we should not do that.
You have to handle the multiple audiences when writing the token.